fastly/observability_custom_dashboard.go

```diff
@@ -0,0 +1,275 @@
+	Title string `json:"title"`
+	// Visualization describes the way the DashboardItem should display data (required)
+	Visualization DashboardVisualization `json:"visualization"`
+}
+
+type DashboardSourceType string
+
+const (
+	SourceTypeStatsEdge   = "stats.edge"
+	SourceTypeStatsDomain = "stats.domain"
+	SourceTypeStatsOrigin = "stats.origin"
+)
+
+// DashboardDataSource describes the data to display in a DashboardItem
+type DashboardDataSource struct {
+	// Config describes configuration options for the selected data source (required)
+	Config DashboardSourceConfig `json:"config"`
+	// Type is the source of the data to display (required)
+	Type DashboardSourceType `json:"type"`
+}
+
+type DashboardSourceConfig struct {
+	// Metrics is the list metrics to visualize (required)
+	// Valid options are defined by the selected SourceType. See https://www.fastly.com/documentation/reference/api/observability/custom-dashboards/#data-source
+	Metrics []string `json:"metrics"`
+}
+
+type VisualizationType string
+
+const VisualizationTypeChart VisualizationType = "chart"
+
+type DashboardVisualization struct {
+	// Config describes configuration options for the given visualization (required)
+	Config VisualizationConfig `json:"config"`
+	// Type is type of visualization to display. Currently only "chart" is supported (required)
+	Type VisualizationType `json:"type"`
+}
+
+type PlotType string
+
+const (
+	PlotTypeLine         PlotType = "line"
+	PlotTypeBar          PlotType = "bar"
+	PlotTypeDonut        PlotType = "donut"
+	PlotTypeSingleMetric PlotType = "single-metric"
+)
+
+type VisualizationFormat string
+
+const (
+	VisualizationFormatNumber       VisualizationFormat = "number"
+	VisualizationFormatBytes        VisualizationFormat = "bytes"
+	VisualizationFormatPercent      VisualizationFormat = "percent"
+	VisualizationFormatRequests     VisualizationFormat = "requests"
+	VisualizationFormatResponses    VisualizationFormat = "responses"
+	VisualizationFormatSeconds      VisualizationFormat = "seconds"
+	VisualizationFormatMilliseconds VisualizationFormat = "milliseconds"
+	VisualizationFormatRatio        VisualizationFormat = "ratio"
+	VisualizationFormatBitrate      VisualizationFormat = "bitrate"
+)
+
+type CalculationMethod string
+
+const (
+	CalculationMethodAvg    CalculationMethod = "avg"
+	CalculationMethodSum    CalculationMethod = "sum"
+	CalculationMethodMin    CalculationMethod = "min"
+	CalculationMethodMax    CalculationMethod = "max"
+	CalculationMethodLatest CalculationMethod = "latest"
+)
+
+type VisualizationConfig struct {
+	// CalculationMethod is the aggregation function to apply to the dataset (optional)
+	CalculationMethod *CalculationMethod `json:"calculation_method,omitempty"`
+	// Format indicates the unit used to format the data (optional, default: number)
+	Format *VisualizationFormat `json:"format,omitempty"`
+	// PlotType is the type of chart to display (required)
+	PlotType PlotType `json:"plot_type"`
+}
+
```

Comment:
@kpfleming Do you think it's worth doing something like this to abstract-away the schema of the nested `DashboardItem` resources? If we supply a constructor, then the user doesn't need to know/care as much about how the data is modeled.

```suggestion
type DashboardItemOption interface {
	apply(*DashboardItem)
}
type optionFunc func(*DashboardItem)
func (f optionFunc) apply(di *DashboardItem) {
	f(di)
}

func WithTitle(title string) DashboardItemOption {
	return optionFunc(func(di *DashboardItem) {
		di.Title = title
	})
}
func WithSubtitle(subtitle string) DashboardItemOption {
	return optionFunc(func(di *DashboardItem) {
		di.Subtitle = subtitle
	})
}
func WithSpan(span uint8) DashboardItemOption {
	return optionFunc(func(di *DashboardItem) {
		di.Span = span
	})
}
func WithCalculationMethod(calculationMethod CalculationMethod) DashboardItemOption {
	return optionFunc(func(di *DashboardItem) {
		di.Visualization.Config.CalculationMethod = &calculationMethod
	})
}
func WithFormat(format VisualizationFormat) DashboardItemOption {
	return optionFunc(func(di *DashboardItem) {
		di.Visualization.Config.Format = &format
	})
}

func NewDashboardItem(sourceType DashboardSourceType, metrics []string, plotType PlotType, options ...DashboardItemOption) DashboardItem {
	di := DashboardItem{
		DataSource: DashboardDataSource{
			Type: sourceType,
			Config: DashboardSourceConfig{
				Metrics: metrics,
			},
		},
		Visualization: DashboardVisualization{
			Type: VisualizationTypeChart,
			Config: VisualizationConfig{
				PlotType: plotType,
			},
		},
	}

	for _, o := range options {
		o.apply(&di)
	}

	return di
}

```

Reply:
That might be useful... you could also consider implementing this using the Builder pattern, so that the invocation of `NewDashboardItem` doesn't get increasingly longer as more and more options are added.